### PR TITLE
Token Locking

### LINKED ISSUE #217: Token Locking
Create token locking mechanism that supports _per-user_ locking functionality. It should be a level higher than the colony, meaning that multiple colonies should be able to lock the same token, as it is possible that multiple colonies have the same native token, which was made possible by BYOT functionality. It should also be universal, meaning that in can be used for variety of purposes like reward payouts, token-weighed voting etc.

### PR DESCRIPTION
**Pull request by filiplazovic**
Closes #217 

I think we are familiar with locking functionality with counters. Additional changes include:
- Token lock counters will replace reward payout counters
- Waiving reward payouts will be done on token locking contract by increasing token lock count
- A colony can lock everyones tokens, except for addresses that have 0 deposited
- When address that have 0 deposited balance deposits the tokens, their token lock count will be set to total lock count

I know that tests for reward payouts need refactoring, it will be fixed in separate PR

### CODE REVIEW

File: contracts/ColonyFunding.sol [hunk 1]
```diff
@@ -166,44 +167,38 @@ contract ColonyFunding is ColonyStorage, DSMath {
...
-  function getGlobalRewardPayoutCount() public view returns (uint256) {
-    return globalRewardPayoutCount;
-  }
-
-  function getUserRewardPayoutCount(address _user) public view returns (uint256) {
-    return userRewardPayoutCount[_user];
-  }
-
   function startNextRewardPayout(address _token) public auth {
+    ITokenLocking tokenLocking = ITokenLocking(IColonyNetwork(colonyNetworkAddress).getTokenLocking());
```
Comment (filiplazovic) [inline contracts/ColonyFunding.sol]: I guess whoever controls the colony network will be able to change token locking address without colony knowing, so i will change that

File: contracts/Authority.sol [hunk 2]
```diff
@@ -26,55 +26,50 @@ contract Authority is DSRoles {
...
+    setAdminRoleCapability(colony, "setAdminRole(address)");
+    setOwnerRoleCapability(colony, "setAdminRole(address)");
+  }
+
+  function setOwnerRoleCapability(address colony, bytes sig) private {
+    bytes4 functionSig = bytes4(keccak256(sig));
+    setRoleCapability(ownerRole, colony, functionSig, true);
+  }
+
+  function setAdminRoleCapability(address colony, bytes sig) private {
```
Comment (elenadimitrova) [inline contracts/Authority.sol]: Sorry what does this have to do with token locking?

File: contracts/TokenLocking.sol [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+  function getTotalLockCount(address _token) public view returns (uint256) {
+    return totalLockCount[_token];
+  }
+
+  function getUserLockCount(address _token, address _user) public view returns (uint256) {
+    return userLocks[_token][_user].lockCount;
+  }
+
+  function getUserDepositedBalance(address _token, address _user) public view returns (uint256) {
```
Comment (elenadimitrova) [inline contracts/TokenLocking.sol]: Can we merge these two functions to return the `Lock` struct?

File: contracts/Colony.sol [hunk 4]
```diff
@@ -59,6 +59,14 @@ contract Colony is ColonyStorage, PatriciaTreeProofs {
     return token;
   }
 
+  function updateTokenLocking() public auth {
+    tokenLocking = ITokenLocking(IColonyNetwork(colonyNetworkAddress).getTokenLocking());
+  }
```
[the remark this answers is not in the document]
Comment (filiplazovic) [reply contracts/Colony.sol]: Upgrading token locking contract is one thing, but i have another concern.
Colony network owners are able to add function on `ColonyNetwork` contract that can completely change address of token locking contract

File: contracts/TokenLocking.sol (hunk 3 above)
Comment (filiplazovic) [reply contracts/TokenLocking.sol]: Right now these two functions are not used anywhere together, so i would like it to stay the as it is.

File: contracts/ColonyFunding.sol [hunk 5]
```diff
@@ -166,44 +167,37 @@ contract ColonyFunding is ColonyStorage, DSMath {
...
 
-    emit RewardPayoutCycleStarted(globalRewardPayoutCount);
+    emit RewardPayoutCycleStarted(totalLockCount);
   }
 
   function claimRewardPayout(uint256 _payoutId, uint256[7] _squareRoots, uint256 _userReputation, uint256 _totalReputation) public {
     RewardPayoutCycle memory payout = rewardPayoutCycles[_payoutId];
+    // Checking if payout is active
     require(block.timestamp - payout.blockTimestamp <= 60 days, "colony-reward-payout-not-active");
-    require(_payoutId - userRewardPayoutCount[msg.sender] == 1, "colony-reward-payout-bad-id");
```
[the remark this answers is not in the document]
Comment (filiplazovic) [reply contracts/ColonyFunding.sol]: No, will add it.
Since we only allow claiming next in 'queue', we could also not pass `_payoutId` and just take current users token lock count and add `1`. But i guess its not significantly different that this

**Comment by filiplazovic** [comment]
I will log an issue and replace staking implementation with this in another PR.
Authority contract changes will be moved to another branch, since we will have `stack too deep` error as soon as we add one more variable.

File: test/colony-funding.js [hunk 6]
```diff
@@ -842,6 +826,158 @@ contract("Colony Funding", addresses => {
...
+      ({ logs } = await colony2.startNextRewardPayout(otherToken.address));
+      const payoutId2 = logs[0].args.id;
+
+      await colony1.claimRewardPayout(payoutId1.toString(), initialSquareRoots1, userReputation1.toString(), totalReputation.toString(), {
+        from: userAddress1
+      });
+
+      await colony2.claimRewardPayout(payoutId2.toString(), initialSquareRoots1, userReputation1.toString(), totalReputation.toString(), {
+        from: userAddress1
+      });
```
Comment (elenadimitrova) [inline test/colony-funding.js]: Can we check the user balance to confirm success?

File: test/colony-funding.js [hunk 7]
```diff
@@ -842,6 +826,158 @@ contract("Colony Funding", addresses => {
...
+      ({ colonyAddress } = logs[0].args);
+      const colony2 = IColony.at(colonyAddress);
+
+      // Giving both colonies the capability to call `mint` function
+      const adminRole = 1;
+      const newRoles = await DSRoles.new();
+      await newRoles.setUserRole(colony1.address, adminRole, true);
+      await newRoles.setUserRole(colony2.address, adminRole, true);
+      await newRoles.setRoleCapability(adminRole, newToken.address, sha3("mint(uint256)").slice(0, 10), true);
+      await newToken.setAuthority(newRoles.address);
```
Comment (elenadimitrova) [inline test/colony-funding.js]: Can't we use `await newToken.setOwner(colony1 .address);` to simplify the above?

Comment (filiplazovic) [reply test/colony-funding.js]: To replace it with all of this above? Both `colony1` and `colony2` need to have permission to call `mint` function on the token contract

Comment (elenadimitrova) [reply test/colony-funding.js]: ah yes, there can only be one owner..
